### ISSUE TITLE
[inductor] `torch.var_mean` / `torch.std_mean` on an empty tensor return mean `0.` instead of `nan` under `torch.compile`

### ISSUE THREAD

**Issue by cuiliaomei-beep** [pytorch/pytorch#197108]
### 🐛 Describe the bug

**Summary.** `torch.var_mean` and `torch.std_mean` on an empty tensor do not agree between eager and Inductor: eager and `aot_eager` return `(nan, nan)`, while Inductor returns `(nan, 0.)` for the mean component (CPU and CUDA, torch 2.14.0 and nightly, for correction 0 and 1, float32 and float16, shapes `(0,)` and `(0, 1)`). A user who reduces an occasionally empty batch gets a finite mean that hides the empty input. Because the mean of zero elements is decided by the codegen's reduction identity rather than by the eager definition, the two paths disagree only in this corner.

For a zero-element input, eager `torch.var_mean` / `torch.std_mean` return `(nan, nan)` (with the usual "degrees of freedom is <= 0" warning). Under `torch.compile(backend="inductor")` the result is `(nan, 0.)`: the **mean** component is `0.` instead of `nan`. `backend="eager"` and `backend="aot_eager"` match eager, so the deviation enters at Inductor lowering/codegen. `torch.mean`, `torch.var` and `torch.std` on the same input are correct under Inductor; only the fused `var_mean` / `std_mean` are affected.

Reproduces for: `var_mean` and `std_mean`; `correction=1` (default) and `correction=0` / `unbiased=False`; float32 and float16; shapes `(0,)` and `(0, 1)`; `dynamic=True` and `False`. Reducing an empty dimension with `dim=` on a `(0, 3)` input gives `nan` correctly in both — only the full reduction to a scalar is affected.

```python
import torch

x = torch.randn(0)

def f(a):
    return torch.var_mean(a)

print("eager    :", f(x))                                        # (tensor(nan), tensor(nan))
torch._dynamo.reset()
print("aot_eager:", torch.compile(f, backend="aot_eager")(x))   # (tensor(nan), tensor(nan))
torch._dynamo.reset()
print("inductor :", torch.compile(f, backend="inductor")(x))    # (tensor(nan), tensor(0.))   <-- mean differs
torch._dynamo.reset()
print("mean only:", torch.compile(lambda a: a.mean(), backend="inductor")(x))   # tensor(nan)  (correct)
torch._dynamo.reset()
print("std_mean :", torch.compile(lambda a: torch.std_mean(a), backend="inductor")(x))   # (tensor(nan), tensor(0.))
```

Expected: `(tensor(nan), tensor(nan))` — the mean of zero elements has no value (same as eager and as `torch.mean` under Inductor).

Actual: `(tensor(nan), tensor(0.))`.

Not CPU- or Windows-specific: on Linux (Kaggle, torch 2.10.0+cu128) both the CPU and the **Tesla T4** Inductor paths return `(nan, 0.)` for `var_mean` and `std_mean` on an empty input while eager and aot_eager return `(nan, nan)`.

Still present on nightly torch 2.15.0.dev20260911+cu130 (nightly, git fb8ff0f1, Kaggle: Tesla T4 + Linux CPU): inductor `(nan, 0.)` vs eager/aot_eager `(nan, nan)` for both `var_mean` and `std_mean`, CPU and CUDA.

The same wrong mean is produced through the second compilation front end: `torch.export(...)` +
`torch._inductor.aoti_compile_and_package` + `aoti_load_package` returns `(nan, 0.)` for `var_mean` and `std_mean` on an
empty input as well (2.14.0+cpu), so the defect is in the shared Inductor lowering, not in Dynamo/AOTAutograd.

Notes:
- Not the case of #113167 (scalar input with an invalid `correction`; closed 2023-11): here the input has zero elements, `correction` is the default, and the wrong component is the mean.
- Eager's NaN over an empty dimension is the intended semantics for reductions without an identity (#61901; see also #191229 for `median`).
- Found by a differential harness that compares eager / `backend="eager"` / `aot_eager` / `inductor` and flags a result only when it disagrees with a float64 reference; reproduced 2/2 with a fixed seed and under four independent contexts (fp32 / fp16 / requires_grad / dynamic=True).

### Error logs

No error is raised; the values differ as shown above.

### Versions

```
PyTorch version: 2.14.0+cpu
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: Microsoft Windows 11 Pro for Workstations (10.0.26200 64-bit)
GCC version: Could not collect
Clang version: Could not collect
CMake version: Could not collect
Libc version: N/A
MSVC: Microsoft (R) C/C++ Optimizing Compiler Version 19.44.35228 for x64 (used by Inductor's C++ backend)

Python version: 3.14.7 (tags/v3.14.7:823f032, Aug  5 2026, 10:51:32) [MSC v.1944 64 bit (AMD64)] (64-bit runtime)
Python platform: Windows-11-10.0.26200-SP0
Is CUDA available: False
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Is XPU available: False
Is XNNPACK available: False

CPU:
Name: 11th Gen Intel(R) Core(TM) i9-11900T @ 1.50GHz
Manufacturer: GenuineIntel
Architecture: 9

Versions of relevant libraries:
[pip3] numpy==2.5.2
[pip3] torch==2.14.0+cpu
[pip3] torchvision==0.29.0+cpu
[conda] Could not collect
```

cc @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @aakhundov @coconutruben @jataylo

**Comment by he-yufeng** [comment]
Root-caused this to the Welford lowering path in `torch/_inductor/lowering.py`, and the scope table in the report falls out of the path selection logic exactly.

Why only the scalar full reduction is affected: `use_two_step_variance()` returns False exactly when the output is a scalar (`sympy_product(ranges) == 1` early-returns False), so `var_mean` on a `(0,)` input goes through `var_mean_welford_()`, while reducing an empty dim on a `(0, 3)` input takes the two-step `var_mean_sum_()` path.

In `var_mean_welford_()` the mean output is the raw Welford accumulator from `ir.WelfordReduction.create()`. On an empty reduction the accumulator never leaves its identity state (mean=0, m2=0, count=0), and unlike the two-step path there is no final divide by the element count, so the mean comes out as 0. The var component is only correct by accident here: `scale_fn` computes `m2 / max(0, N - correction)` = 0/0 = nan. The two-step path gets the empty mean right because the standalone `mean()` lowering divides sum=0 by denom=0 and yields nan, which is also why `torch.mean` alone is fine.

`aten.std_mean` decomposes into `torch.var_mean` (`torch/_refs/__init__.py`), so one fix in the same lowering covers both ops. The path selection is device-independent for the scalar case, which matches the report reproducing on CPU and CUDA.

Proposed fix: in `var_mean_welford_()`, when `return_mean` is set, mask the realized mean with `where(rnumel == 0, nan, mean)`. For static shapes `rnumel` is a constant so the select folds away for every non-empty case; for dynamic shapes it is one select on a scalar output. I have the patch and a regression test sketched (empty `(0,)` and `(0, 1)`, correction 0 and 1, `var_mean` plus `std_mean`, and the non-empty Welford path asserted unchanged). Happy to put up the PR if a maintainer can mark this actionable.
